keyboards/mlego/m20/m20.c

```diff
@@ -0,0 +1,150 @@
+
+#endif
+
+#ifdef OLED_ENABLE
+
+static uint32_t oled_logo_timer = 0;
+static bool clear_logo = true;
+static const char PROGMEM m65_logo[] = {
+    0x91, 0x92, 0x93, 0x94, 0x95, 0x96, 0x97, 0x98, 0x99, 0xa0, 0xa1, 0xa2, 0xa3, 0xa4, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5,
+    0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89, 0x8a, 0x8b, 0x8c, 0x8d, 0x8e, 0x8f, 0x90, 0x91, 0x92, 0x93, 0x94,
+    0xa0, 0xa1, 0xa2, 0xa3, 0xa4, 0xa5, 0xa6, 0xa7, 0xa8, 0xa9, 0xaa, 0xab, 0xac, 0xad, 0xae, 0xaf, 0xb0, 0xb1, 0xb2, 0xb3, 0xb4,
+    0xc0, 0xc1, 0xc2, 0xc3, 0xc4, 0xc5, 0xc6, 0xc7, 0xc8, 0xc9, 0xca, 0xcb, 0xcc, 0xcd, 0xce, 0xcf, 0xd0, 0xd1, 0xd2, 0xd3, 0xd4,
+    0};
+
+#endif
+
+#ifdef RGBLIGHT_ENABLE
+
+void set_rgb_layers(layer_state_t state){
+
+    rgblight_set_layer_state(0, layer_state_cmp(state, _NUM));
+    rgblight_set_layer_state(1, layer_state_cmp(state, _LWR));
+    rgblight_set_layer_state(2, layer_state_cmp(state, _RSE));
+    rgblight_set_layer_state(3, layer_state_cmp(state, _ADJ));
+
+}
+
+void set_default_rgb_layers(layer_state_t state){
+    rgblight_set_layer_state(0, layer_state_cmp(state, _NUM));
+}
+
+const rgblight_segment_t * const* my_rgb(void){
+    return my_rgb_layers;
+}
+
+#endif
+
+void set_led_toggle(const uint8_t layer, const bool state){
+
+    switch (layer) {
+        case _LWR:
+          toggle_lwr = state;
+          break;
+        case _RSE:
+          toggle_rse = state;
+          break;
+        default:
+          break;
+    }
+}
+
+void toggle_leds(void){
+
+    led_lwr(toggle_lwr);
+    led_rse(toggle_rse);
+      if (layer_state_is(_ADJ)) {
+        led_lwr(true);
+        led_rse(true);
+    }
+
+}
+
+#ifdef OLED_ENABLE
+
+void init_timer(void){
+   oled_logo_timer = timer_read32();
+};
+
+void user_oled_magic(void) {
+    // Host Keyboard Layer Status
+    oled_write_P(PSTR("Layer: "), false);
+
+    switch (get_highest_layer(layer_state)) {
+        case _NUM:
+            oled_write_P(PSTR("Default\n"), false);
+            break;
+        case _LWR:
+            oled_write_P(PSTR("Lower\n"), false);
+            break;
+        case _RSE:
+            oled_write_P(PSTR("Raise\n"), false);
+            break;
+        case _ADJ:
+            oled_write_P(PSTR("ADJ\n"), false);
+            break;
+        default:
+            // Or use the write_ln shortcut over adding '\n' to the end of your string
+            oled_write_ln_P(PSTR("Undefined"), false);
+    }
+
+    // Host Keyboard LED Status
+    led_t led_state = host_keyboard_led_state();
+    oled_write_P(led_state.num_lock ? PSTR("Lower ") : PSTR("    "), false);
+    oled_write_P(led_state.scroll_lock ? PSTR("Raise ") : PSTR("    "), false);
+#ifdef WPM_ENABLE
+    oled_write_P(PSTR("\nwpm: "), false);
+    uint8_t wpm = get_current_wpm();
+    oled_write_P(wpm != 0 ? get_u8_str(wpm,' ') : PSTR("   "), false);
+#endif
+}
```

Comment:
This looks like it's all keymap-specific code? Anything that depends on things like layers or that not every user may want should be moved to a custom keymap.

Reply:
I am in two minds about this and over years I was given various suggestions. I can move it all in the keymap, but if I want to share the code between keymaps where do I put it?

Reply:
happy to move it in the userspace if that is ok

Reply:
done, hope properly